File: projections/dotnet/foundation/hlp.foundation.rule-runtime/Evaluation/DateMath.cs
using System.Globalization;

namespace Harborline.Foundation.RuleEngine.Evaluation;

/// <summary>
/// Deterministic, timezone-free calendar arithmetic for the <c>date.*</c>
/// operators (SPINE-1 design §1.4). Dates are <c>YYYY-MM-DD</c> in the proleptic
/// Gregorian calendar (UTC, date-only — DST is structurally excluded). Uses
/// Howard Hinnant's days-from-civil algorithm so leap years are exact and the
/// result is <b>byte-identical</b> to the TS tier's port.
/// </summary>
internal static class DateMath
{
    /// <summary>Parses a <c>YYYY-MM-DD</c> date into (year, month, day); throws on malformed input.</summary>
    public static (int Y, int M, int D) Parse(string text)
    {
        if (text is null || text.Length < 8) throw new FormatException($"invalid date '{text}'");
        var parts = text.Split('-');
        // Support an optional time suffix by truncating at 'T'.
        if (parts.Length >= 3)
        {
            var dayPart = parts[2];
            int tIdx = dayPart.IndexOf('T');
            if (tIdx >= 0) dayPart = dayPart[..tIdx];
            if (int.TryParse(parts[0], NumberStyles.None, CultureInfo.InvariantCulture, out int y)
                && int.TryParse(parts[1], NumberStyles.None, CultureInfo.InvariantCulture, out int m)
                && int.TryParse(dayPart, NumberStyles.None, CultureInfo.InvariantCulture, out int d)
                && m is >= 1 and <= 12 && d is >= 1 and <= 31)
            {
                return (y, m, d);
            }
        }
        throw new FormatException($"invalid date '{text}'");
    }

    /// <summary>Howard Hinnant's days_from_civil: serial day number (1970-01-01 = 0).</summary>
    public static long EpochDay(int y, int m, int d)
    {
        long yy = m <= 2 ? y - 1 : y;
        long era = (yy >= 0 ? yy : yy - 399) / 400;
        long yoe = yy - era * 400;                                  // [0, 399]
        long doy = (153 * (m > 2 ? m - 3 : m + 9) + 2) / 5 + d - 1; // [0, 365]
        long doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;           // [0, 146096]
        return era * 146097 + doe - 719468;
    }

    /// <summary>Inverse of <see cref="EpochDay"/>: civil date from a serial day number.</summary>
    public static (int Y, int M, int D) FromEpochDay(long z)
    {
        z += 719468;
        long era = (z >= 0 ? z : z - 146096) / 146097;
        long doe = z - era * 146097;                                  // [0, 146096]
        long yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365; // [0, 399]
        long y = yoe + era * 400;
        long doy = doe - (365 * yoe + yoe / 4 - yoe / 100);           // [0, 365]
        long mp = (5 * doy + 2) / 153;                                 // [0, 11]
        long d = doy - (153 * mp + 2) / 5 + 1;                         // [1, 31]
        long m = mp < 10 ? mp + 3 : mp - 9;                            // [1, 12]
        return ((int)(m <= 2 ? y + 1 : y), (int)m, (int)d);
    }

    private static int DaysInMonth(int y, int m)
    {
        int[] lengths = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
        if (m == 2 && IsLeap(y)) return 29;
        return lengths[m - 1];
    }

    private static bool IsLeap(int y) => (y % 4 == 0 && y % 100 != 0) || y % 400 == 0;

    private static string Format(int y, int m, int d)
        => $"{y:D4}-{m:D2}-{d:D2}";

    /// <summary>Today's UTC date (from an injected clock) as <c>YYYY-MM-DD</c>.</summary>
    public static string Today(DateTimeOffset nowUtc)
        => Format(nowUtc.UtcDateTime.Year, nowUtc.UtcDateTime.Month, nowUtc.UtcDateTime.Day);

    /// <summary>Adds <paramref name="n"/> of <paramref name="unit"/> (day|month|year) with day-clamping.</summary>
    public static string Add(string date, long n, string unit)
    {
        var (y, m, d) = Parse(date);
        switch (unit)
        {
            case "day":
                var (ny, nm, nd) = FromEpochDay(EpochDay(y, m, d) + n);
                return Format(ny, nm, nd);
            case "month":
            {
                long total = (long)y * 12 + (m - 1) + n;
                int yr = (int)Math.DivRem(total, 12, out long mo);
                if (mo < 0) { mo += 12; yr -= 1; }
                int month = (int)mo + 1;
                int day = Math.Min(d, DaysInMonth(yr, month));
                return Format(yr, month, day);
            }
            case "year":
            {
                int yr = y + (int)n;
                int day = Math.Min(d, DaysInMonth(yr, m));
                return Format(yr, m, day);
            }
            default:
                throw new FormatException($"unknown date unit '{unit}'");
        }
    }

    /// <summary>Difference <c>a - b</c> in whole days (v1 supports the <c>day</c> unit).</summary>
    public static long DiffDays(string a, string b)
    {
        var (ay, am, ad) = Parse(a);
        var (by, bm, bd) = Parse(b);
        return EpochDay(ay, am, ad) - EpochDay(by, bm, bd);
    }
}

File: projections/typescript/foundation/hlp.foundation.rule-runtime/src/date-math.ts
/**
 * Deterministic, timezone-free calendar arithmetic for the `date.*` operators
 * (SPINE-1 §1.4). Dates are `YYYY-MM-DD` (proleptic Gregorian, UTC, date-only —
 * DST structurally excluded). Howard Hinnant's days-from-civil algorithm, so leap
 * years are exact and the result is byte-identical to the .NET `DateMath` port.
 */

function parse(text: string): [number, number, number] {
  const parts = (text ?? '').split('-')
  if (parts.length >= 3) {
    let dayPart = parts[2]
    const tIdx = dayPart.indexOf('T')
    if (tIdx >= 0) dayPart = dayPart.slice(0, tIdx)
    const y = Number(parts[0])
    const m = Number(parts[1])
    const d = Number(dayPart)
    if (
      Number.isInteger(y) && Number.isInteger(m) && Number.isInteger(d) &&
      m >= 1 && m <= 12 && d >= 1 && d <= 31
    ) {
      return [y, m, d]
    }
  }
  throw new Error(`invalid date '${text}'`)
}

export function epochDay(y: number, m: number, d: number): number {
  const yy = m <= 2 ? y - 1 : y
  const era = Math.floor((yy >= 0 ? yy : yy - 399) / 400)
  const yoe = yy - era * 400
  const doy = Math.floor((153 * (m > 2 ? m - 3 : m + 9) + 2) / 5) + d - 1
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy
  return era * 146097 + doe - 719468
}

function fromEpochDay(z0: number): [number, number, number] {
  const z = z0 + 719468
  const era = Math.floor((z >= 0 ? z : z - 146096) / 146097)
  const doe = z - era * 146097
  const yoe = Math.floor((doe - Math.floor(doe / 1460) + Math.floor(doe / 36524) - Math.floor(doe / 146096)) / 365)
  const y = yoe + era * 400
  const doy = doe - (365 * yoe + Math.floor(yoe / 4) - Math.floor(yoe / 100))
  const mp = Math.floor((5 * doy + 2) / 153)
  const d = doy - Math.floor((153 * mp + 2) / 5) + 1
  const m = mp < 10 ? mp + 3 : mp - 9
  return [m <= 2 ? y + 1 : y, m, d]
}

function isLeap(y: number): boolean {
  return (y % 4 === 0 && y % 100 !== 0) || y % 400 === 0
}

function daysInMonth(y: number, m: number): number {
  const lengths = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
  if (m === 2 && isLeap(y)) return 29
  return lengths[m - 1]
}

function fmt(y: number, m: number, d: number): string {
  const p = (n: number, w: number) => String(n).padStart(w, '0')
  return `${p(y, 4)}-${p(m, 2)}-${p(d, 2)}`
}

export function today(nowUtc: Date): string {
  return fmt(nowUtc.getUTCFullYear(), nowUtc.getUTCMonth() + 1, nowUtc.getUTCDate())
}

export function dateAdd(date: string, n: number, unit: string): string {
  const [y, m, d] = parse(date)
  switch (unit) {
    case 'day': {
      const [ny, nm, nd] = fromEpochDay(epochDay(y, m, d) + n)
      return fmt(ny, nm, nd)
    }
    case 'month': {
      const total = y * 12 + (m - 1) + n
      let yr = Math.floor(total / 12)
      let mo = total % 12
      if (mo < 0) {
        mo += 12
        yr -= 1
      }
      const month = mo + 1
      const day = Math.min(d, daysInMonth(yr, month))
      return fmt(yr, month, day)
    }
    case 'year': {
      const yr = y + n
      const day = Math.min(d, daysInMonth(yr, m))
      return fmt(yr, m, day)
    }
    default:
      throw new Error(`unknown date unit '${unit}'`)
  }
}

export function dateDiffDays(a: string, b: string): number {
  const [ay, am, ad] = parse(a)
  const [by, bm, bd] = parse(b)
  return epochDay(ay, am, ad) - epochDay(by, bm, bd)
}
